### PR TITLE
[MID-85] Sub-navigation menu to allow user to see specialists based on the district

### PR DESCRIPTION
**Pull request by watchlar2000**
[MID-85](https://trello.com/c/VAfWf2bT/85-%D1%8E%D0%B7%D0%B5%D1%80-%D0%BF%D0%BE%D1%88%D1%83%D0%BA-%D0%BF%D0%BE-%D1%80%D0%B0%D0%B9%D0%BE%D0%BD%D0%B0%D0%BC-%D0%B7-%D0%B3%D0%BE%D0%BB%D0%BE%D0%B2%D0%BD%D0%BE%D1%97-%D1%81%D1%82%D0%BE%D1%80%D1%96%D0%BD%D0%BA%D0%B8)

image

### CODE REVIEW

File: src/app/_components/DistrictSearch/DistrictList.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+'use client';
+
+import { Tick } from '@icons/index';
+import Link from 'next/link';
+import PropTypes from 'prop-types';
+import { useState } from 'react';
+import { Swiper, SwiperSlide } from 'swiper/react';
+
+import 'swiper/css';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictList.js]: 🟢 shouldn't that be somewhere closer to root css? 🤔

File: src/app/_components/DistrictSearch/DistrictList.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+'use client';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictList.js]: 🟠 since its already folder for 'DistrictSearch' - could 'DistrictList' be shorted to 'List' ?

File: src/app/_components/DistrictSearch/DistrictSearch.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import PropTypes from 'prop-types';
+import { cn } from '@utils/cn';
+import { Heading } from '@/app/_components';
+import { getDistrictsList } from '@/app/_components/DistrictSearch/actions';
+import { DistrictList } from '@/app/_components/DistrictSearch/DistrictList';
+
+export async function DistrictSearch({ className }) {
+  const optionsList = await getDistrictsList();
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictSearch.js]: 🟢 do you think its worth creating helper file for 3 lines of code? 🤔

File: src/app/_components/DistrictSearch/DistrictList.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  };
+
+  return (
+    <Swiper slidesPerView="auto" spaceBetween={15} wrapperClass={className}>
+      {list.map(({ id, name }, index) => {
+        const isSelected = index === selected;
+        const hasIcon = isSelected ? <Tick /> : null;
+
+        return (
+          <SwiperSlide key={id} onClick={() => handleClick(index)} className="fit-co">
```
Comment (ZakharovMaxim) [inline src/app/_components/DistrictSearch/DistrictList.js]: fit-co -> fit-content?

File: src/app/_components/DistrictSearch/DistrictList.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            <Link
+              href={`/specialist?district=${id}`}
+              className={isSelected ? 'pointer-events-none cursor-none' : ''}
+              tabIndex={-1}
+            >
+              <PillButton
+                variant="transparent"
+                colorVariant="orange"
+                icon={hasIcon}
+                className={isSelected ? 'border-secondary-300 bg-secondary-300 font-semibold' : ''}
```
Comment (ZakharovMaxim) [inline src/app/_components/DistrictSearch/DistrictList.js]: Use clsx package

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 1 above)
Comment (ZakharovMaxim) [inline src/app/_components/DistrictSearch/DistrictList.js]: You can create a component named Slider, which will import Swiper and swiper.css and reexport Swiper and SwiperSlide.

File: src/app/_components/DistrictSearch/DistrictList.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    <Swiper slidesPerView="auto" spaceBetween={15} wrapperClass={className}>
+      {list.map(({ id, name }, index) => {
+        const isSelected = index === selected;
+        const hasIcon = isSelected ? <Tick /> : null;
+
+        return (
+          <SwiperSlide key={id} onClick={() => handleClick(index)} className="fit-co">
+            <Link
+              href={`/specialist?district=${id}`}
+              className={isSelected ? 'pointer-events-none cursor-none' : ''}
```
Comment (ZakharovMaxim) [inline src/app/_components/DistrictSearch/DistrictList.js]: Use clsx package

File: src/app/_components/DistrictSearch/DistrictSearch.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+  return (
+    <section className={cn('mx-auto w-full max-w-[900px] px-4 lg:px-0', className)}>
+      <Heading type="h3" className="text-p4 font-bold uppercase text-primary-600">
+        Райони міста Львова
+      </Heading>
+      <DistrictList list={optionsList} className="mt-4 w-[100%]" />
```
Comment (ZakharovMaxim) [inline src/app/_components/DistrictSearch/DistrictSearch.js]: use w-full

File: src/app/_components/PillButton/style.js [hunk 8]
```diff
@@ -42,6 +42,15 @@ export const buttonColorVariant = {
       disabled,
     },
   },
+  transparent: {
+    orange: {
+      regular: 'bg-transparent border-[1px] border-solid border-gray-700 text-p4 text-gray-900 font-normal',
```
Comment (ZakharovMaxim) [inline src/app/_components/PillButton/style.js]: I think border-[1px] is the same as just border

**Review by ZakharovMaxim — CHANGES_REQUESTED** [review]

File: src/app/_components/DistrictSearch/actions.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+'use server';
+
+import { prisma } from '@/lib/db';
+
+export const getDistrictsList = async () => {
+  const districtsList = await prisma.district.findMany({
+    orderBy: { name: 'desc' },
+  });
```
Comment (epaminond) [inline src/app/_components/DistrictSearch/actions.js]: 1. Is it possible that you may have large amount of districts? If so, does it make sense to limit search?
2. It seems you are relying on sorting districts on the backend to display sorted list on the frontend. Would it make sense to make sure it's sorted properly on the frontend instead?

File: src/app/_components/DistrictSearch/actions.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+  const defaultOption = { id: 'all-districts', name: 'Усі' };
+  return [defaultOption, ...districtsList];
```
Comment (epaminond) [inline src/app/_components/DistrictSearch/actions.js]: Is this `defaultOption` just a frontend abstraction? If so, I'd recommend to keep it just on the frontend so that it's "loosely-coupled" (best-practice from architectural perspective).

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 3 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: getDisctrictList is a function which will be executed on the server. That's why I put it into a separate file.

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 4 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: it shouldn't be there at all, removed it

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 5 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: ✅

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 6 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: ✅

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 7 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: ✅

File: src/app/_components/PillButton/style.js (hunk 8 above)
Comment (watchlar2000) [reply src/app/_components/PillButton/style.js]: ✅

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 2 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: Changed name to "List"

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 1 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: Created a Slider component. Could you please take a look at it to see whether I understood your suggestion correctly?

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 3 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: 'page.js' will also be executed on server

File: src/app/_components/DistrictSearch/actions.js (hunk 9 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/actions.js]: just out of curiosity - why filtering on frontend rather than on backend?

Comment (epaminond) [reply src/app/_components/DistrictSearch/actions.js]: Again loose-coupling: order seems to be a frontend-feature, not backend.
Unless you implement pagination or some other order-related logic, of course.

Tight coupling leads to errors. E.g. at some point you decide to store response in frontend-storage, or cache it. And then add some other results to same storage - that would affect order.
And most likely at that point you won't think about necessity to change smth about ordering in components that use it.

Comment (watchlar2000) [reply src/app/_components/DistrictSearch/actions.js]: 1. It's all about Lviv districts, so I'd say that the amount will stay the same for quite a long time. However, in case the amount were bigger, what would be the way to limit search? Pagination? Because all the districts should be available for user to pick from.

2. I though it would be a good idea to have server do all the work including sorting out the districts. What are the benefits of doing it on the front end?

File: src/app/_components/DistrictSearch/actions.js (hunk 10 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/actions.js]: Thanks! Yeah, it could be considered as a frontend abstraction and is just an option selected by default when user visits the main page.

File: src/app/_components/DistrictSearch/actions.js (hunk 9 above)
Comment (epaminond) [reply src/app/_components/DistrictSearch/actions.js]: 1. OK, we can leave it this way for now. If we have them too many, maybe we'd need to implement smth like dropdown search that makes requests based on input.
2. Just answered above. That's also about loose coupling: it would definitely work but you might run into issues long-term if you follow this approach.

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 1 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: But why?

File: src/app/_components/DistrictSearch/DistrictSearch.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import PropTypes from 'prop-types';
+import { cn } from '@utils/cn';
+import { Heading } from '@/app/_components';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictSearch.js]: 🟠 import it directly? @components/Typography

File: src/app/_components/DistrictSearch/DistrictSearch.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import PropTypes from 'prop-types';
+import { cn } from '@utils/cn';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictSearch.js]: 🟢 it is also accessible in @utils i think

File: src/app/_components/DistrictSearch/DistrictSearch.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import PropTypes from 'prop-types';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/DistrictSearch.js]: 🟢 maybe this file could also be renamed to 'component' or something? to avoid duplications DistrictSearch/DistrictSearch.js

File: src/app/_components/DistrictSearch/List.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+'use client';
+
+import { Tick } from '@icons/index';
+import Link from 'next/link';
+import PropTypes from 'prop-types';
+import { useState } from 'react';
+import { SwiperSlide } from 'swiper/react';
+import { PillButton, Slider } from '@components';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/List.js]: 🟠 by convention it should be splited into direct imports

File: src/app/_components/DistrictSearch/List.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+import { clsx } from 'clsx/lite';
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/List.js]: 🟢 isn't there should be 'cn' ?

File: src/app/_components/DistrictSearch/actions.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+// export const getDistrictsList = async () => {
+//   const districtsList = await prisma.district.findMany({
+//     orderBy: { name: 'desc' },
+//   });
+//   const defaultOption = { id: 'all-districts', name: 'Усі' };
+//   return [defaultOption, ...districtsList];
+// };
+
+export const getDistrictsList = async () => await prisma.district.findMany();
```
Comment (alex-popov-tech) [inline src/app/_components/DistrictSearch/actions.js]: 🟠 now that looks even weirder, having whole separate file for just replacing 'prisma' call to 'getDistrictList' call 🤔

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 1 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/DistrictList.js]: to be accessible in some other swiper component in other place? i dunno if it works that way, it just looks like it

File: src/app/_components/DistrictSearch/actions.js (hunk 16 above)
Comment (epaminond) [reply src/app/_components/DistrictSearch/actions.js]: 🟠 By the way it doesn't make sense to put it into `async`-function and `await` results before returning:
- `async` function returns promise
- so in your case you just wrap promise into promise

You'd get same result without extra-wrapping like this:
```javascript
export const getDistrictsList = () => prisma.district.findMany();
```

File: src/app/_components/DistrictSearch/DistrictList.js (hunk 1 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictList.js]: Created a Slider component which will be a wrapper for Swiper.

File: src/app/_components/DistrictSearch/actions.js (hunk 9 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/actions.js]: Got it. Thanks

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 11 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js:3]: You sure it will work?

image

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 12 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: image

File: src/app/_components/DistrictSearch/List.js (hunk 15 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/List.js]: Nope, it shouldn't. Cn combines classes and clsx create a string based on the conditions.

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 13 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: Maybe, though what about other folders? I'd keep it as is and then it will be nice to create a quick ticket to update everything.

image

File: src/app/_components/DistrictSearch/List.js (hunk 14 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/List.js]: Could you please explain?

File: src/app/_components/DistrictSearch/actions.js (hunk 16 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/actions.js]: > 🟠 now that looks even weirder, having whole separate file for just replacing 'prisma' call to 'getDistrictList' call 🤔

It's not about just replacing 'prisma' call. This file has 'use server' directive at the top. The idea is to execute this call on the server.

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 3 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: so what's the question?)

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 11 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: ✅

Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: Though it's a little confusing

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 13 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: both of them are alright in current dev

File: src/app/_components/DistrictSearch/List.js (hunk 14 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/List.js]:
```
import { PillButton } from '@components/PillButton';
import { Slider } from '@components/Slider';
```

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 13 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: Fixed it

File: src/app/_components/DistrictSearch/List.js (hunk 14 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/List.js]: I might be missing something. If we go that way using directed imports, then why do we have an index.js in the components folder?

File: src/app/_components/DistrictSearch/DistrictSearch.js (hunk 11 above)
Comment (alex-popov-tech) [reply src/app/_components/DistrictSearch/DistrictSearch.js]: we've agreed on using direct imports with aliases to avoid accidental importing of server component in client one

File: src/app/_components/DistrictSearch/List.js (hunk 14 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/List.js]: Fixed it

File: src/app/_components/DistrictSearch/actions.js (hunk 16 above)
Comment (watchlar2000) [reply src/app/_components/DistrictSearch/actions.js]: Apparently, based on the approach used on the recent dev branch, this action.js is no longer needed. I removed getDistrictsList function and now this prisma call is executed directly in the components inside DistrictSearch folder.

File: src/app/(app)/page.js [hunk 17]
```diff
@@ -1,4 +1,5 @@
-import { TherapiesSection, FAQSection } from '@components/MainPageSections';
+import { FAQSection, TherapiesSection } from '@components/MainPageSections';
+import { DistrictSearch } from '@components/DistrictSearch';
```
Comment (alex-popov-tech) [inline src/app/(app)/page.js]: 🟠 maybe place this component in 'MainPageSections' like 2 previous?

**Review by alex-popov-tech — APPROVED** [review]

File: src/app/(app)/page.js (hunk 17 above)
Comment (watchlar2000) [reply src/app/(app)/page.js]: Moved it to the MainPageSections folder

File: src/app/_components/Slider.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
+'use client';
+
+import PropTypes from 'prop-types';
+import { Swiper } from 'swiper/react';
+
+import 'swiper/css';
+
+export function Slider({ slidesPerView = 'auto', spaceBetween = 15, children, className }) {
```
Comment (ZakharovMaxim) [inline src/app/_components/Slider.js]: You can keep slidePerView and spaceBetween here if you want to give them default values, but I would also add all others swiper attributes using spread operator, e.g children, className, ...swiperProps }) and later in code <Swiper ...currentAttributes, {...swiperProps} />

File: src/app/_components/Slider.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
+'use client';
+
+import PropTypes from 'prop-types';
+import { Swiper } from 'swiper/react';
```
Comment (ZakharovMaxim) [inline src/app/_components/Slider.js:4]: I would also reexport SwiperSlider named it Slide, so you can import everything from one place and later you can switch to another lib

File: src/app/_components/Slider.js (hunk 18 above)
Comment (watchlar2000) [reply src/app/_components/Slider.js]: Updated Slider component as per your comments

File: src/app/_components/Slider.js (hunk 19 above)
Comment (watchlar2000) [reply src/app/_components/Slider.js:4]: Reexported SwiperSlide

File: docs/CODESTYLE.md [hunk 20]
```diff
@@ -68,17 +70,18 @@ _components
     ├───Create.js // File with SpecialistCreate component
     ├───Edit.js // File with SpecialistEdit component
     ├───Show.js // File with SpecialistShow component
-    ├───List.js // File with SpecialistList component
+    ├───DistrictList.js // File with SpecialistList component
```
Comment (IvanKoshan) [inline docs/CODESTYLE.md]: 🟡 please revert this change if it's accidental. If not explain why it's done

File: src/app/_components/MainPageSections/DistrictList.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+  const [selected, setSelected] = useState(0);
+  const handleClick = index => {
+    setSelected(index);
+  };
+
+  return (
+    <Slider slidesPerView="auto" spaceBetween={15} className={className}>
+      {list.map(({ id, name }, index) => {
+        const isSelected = index === selected;
+        const hasIcon = isSelected ? <Tick /> : null;
```
Comment (IvanKoshan) [inline src/app/_components/MainPageSections/DistrictList.js]: 🟢 This variable name kinda sounds like a boolean value, though it actually isn't. Maybe move this into prop directly, or change name

File: docs/CODESTYLE.md (hunk 20 above)
Comment (watchlar2000) [reply docs/CODESTYLE.md]: It was probably changed automatically, somehow. Fixed it

**Review by IvanKoshan — APPROVED** [review]

**Review by ZakharovMaxim — APPROVED** [review]
